### PR TITLE
Add support for paying by redirection directly to a chosen bank

### LINKED ISSUE #23: direct transfer transactions
* możliwość zdefiniowania, które banki są dostępne bezpośrednio w ramach danej metody płatności
* wyświetlenie danego banku jako metody płatności
* przekierowanie do banku z ominięciem tpay

Ref. https://docs-api.tpay.com/en/payment-methods/pbl/#pay-by-link-transaction

### PR DESCRIPTION
**Pull request by arti0090**
Closes #23

### CODE REVIEW

File: src/Tpay/Provider/TpayApiBankListProvider.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    ) {
+    }
+
+    public function provide(): array
+    {
+        $gateway = $this->payum->getGateway('tpay');
+
+        /** @var GetTpayTransactionsChannels $value */
+        $value = $this->cache->get('tpay_bank_list', function (ItemInterface $item) use ($gateway): GetTpayTransactionsChannels {
+            $item->expiresAfter(new \DateInterval('P1D'));
```
[the remark this answers is not in the document]
Comment (arti0090) [reply src/Tpay/Provider/TpayApiBankListProvider.php:28]: I assumed that new banks are not a thing that happens every hour, hence I choose the 1 day.
Ofc there might be situtation that some payment methods has downtimes but then just the payment will fail somehow.
Still I think its good idea to make it parametrized and configured by admin/env etc.

[the remark this answers is not in the document]
Comment (arti0090) [reply src/Tpay/Provider/TpayApiBankListProvider.php:28]: Saved in https://github.com/orgs/CommerceWeavers/projects/9/views/1?pane=issue&itemId=82030548

**Comment by jakubtobiasz** [comment]
I see there's no API part. Do you plan to add it do this PR or as a separate one? We already have `29` files here, so maybe another PR is a good idea.

File: tests/Application/.env.test [hunk 2]
```diff
@@ -18,5 +18,8 @@ PAYUM_CYPHER_KEY=def00000626f0eff82cfc9ed8bd1f465a22a1ccd4d96e91585a6158975db8dd
 
 ###> commerce-weavers/sylius-tpay-plugin ###
 TPAY_API_URL=<edit_me>
+PANTHER_WEB_SERVER_DIR='./tests/Application/public'
+PANTHER_CHROME_ARGUMENTS='--disable-dev-shm-usage --window-size=1400,900'
+PANTHER_NO_SANDBOX=1
```
Comment (jakubtobiasz) [inline tests/Application/.env.test:23]: Do a rebase, as @coldic3 moved this part to `phpunit.xml.dist`

File: src/Payum/Action/Api/CreateRedirectBasedTransactionAction.php [hunk 3]
```diff
@@ -65,6 +65,9 @@ public function supports($request): bool
 
         $details = $model->getDetails();
 
-        return !isset($details['tpay']['card']) && !isset($details['tpay']['blik_token']);
+        return !isset($details['tpay']['card']) &&
```
Comment (jakubtobiasz) [inline src/Payum/Action/Api/CreateRedirectBasedTransactionAction.php:68]: We should refactor this, as it stinks. Of course, it isn't related to this PR, I'm just leaving it as a note.

File: src/Payum/Action/Api/CreatePayByLinkTransactionAction.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        $response = $this->api->transactions()->createTransaction(
+            $this->createPayByLinkPayloadFactory->createFrom($model, $notifyToken->getTargetUrl(), $localeCode),
+        );
+
+        $details = $model->getDetails();
+        $details['tpay']['transaction_id'] = $response['transactionId'];
+        $details['tpay']['status'] = $response['status'];
+        $details['tpay']['transaction_payment_url'] = $response['transactionPaymentUrl'];
+
+        $model->setDetails($details);
```
Comment (jakubtobiasz) [inline src/Payum/Action/Api/CreatePayByLinkTransactionAction.php:49]: We should use the `PaymentDetails` model:
- https://github.com/CommerceWeavers/SyliusTpayPlugin/blob/main/src/Model/PaymentDetails.php
- https://github.com/CommerceWeavers/SyliusTpayPlugin/blob/e04581ed642074105eb5d7f3042f4122905b7311/src/Payum/Action/Api/CreateBlikLevelZeroTransactionAction.php#L41

File: config/services/tpay.php [hunk 5]
```diff
@@ -96,4 +100,19 @@
         ])
         ->alias(SignatureVerifierInterface::class, 'commerce_weavers_tpay.security.notification.verifier.signature')
     ;
+
+    $services->set('commerce_weavers_tpay.tpay.factory.create_pay_by_link_payment_payload', CreatePayByLinkPayloadFactory::class)
+        ->args([
+            service('commerce_weavers_tpay.tpay.factory.create_redirect_based_payment_payload'),
+        ])
+        ->alias(CreatePayByLinkPayloadFactoryInterface::class, 'commerce_weavers_tpay.factory.create_pay_by_link_payment_payload')
+    ;
```
Comment (jakubtobiasz) [inline config/services/tpay.php:109]: This service should be close to other payload factories

File: translations/messages.pl.yaml [hunk 6]
```diff
@@ -12,6 +12,7 @@ commerce_weavers_sylius_tpay:
                 card: 'Karta (na stronie)'
                 label: 'Typ'
                 redirect: 'Przekierowanie do Tpay'
+                pay_by_link: 'Pay by link'
```
Comment (jakubtobiasz) [inline translations/messages.pl.yaml:15]: It should be rather something like `Przekierowanie do banku`

File: tests/Unit/Payum/Action/Api/NotifyActionTest.php [hunk 7]
```diff
@@ -47,7 +47,7 @@ protected function setUp(): void
         $this->request->getModel()->willReturn($this->model->reveal());
     }
 
-    public function it_supports_only_notify_requests(): void
+    public function test_it_supports_only_notify_requests(): void
```
Comment (jakubtobiasz) [inline tests/Unit/Payum/Action/Api/NotifyActionTest.php:50]: I guess I fixed that some time ago, but good catch 👍🏻

File: src/Payum/Action/Api/CreatePayByLinkTransactionAction.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * @param CreateTransaction $request
+     */
+    public function execute($request): void
+    {
+        /** @var PaymentInterface $model */
+        $model = $request->getModel();
+        $token = $request->getToken();
```
Comment (jakubtobiasz) [inline src/Payum/Action/Api/CreatePayByLinkTransactionAction.php:34]: We should avoid requesting the token to only get the gateway name. It prevents us from resuing it in API. Check the `src/Payum/Action/Api/CreateBlikLevelZeroTransactionAction.php:36` for a reference;

File: src/Payum/Action/Api/CreatePayByLinkTransactionAction.php [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        $notifyToken = $this->notifyTokenFactory->create($model, $token->getGatewayName(), $localeCode);
+
+        $response = $this->api->transactions()->createTransaction(
+            $this->createPayByLinkPayloadFactory->createFrom($model, $notifyToken->getTargetUrl(), $localeCode),
+        );
+
+        $details = $model->getDetails();
+        $details['tpay']['transaction_id'] = $response['transactionId'];
+        $details['tpay']['status'] = $response['status'];
+        $details['tpay']['transaction_payment_url'] = $response['transactionPaymentUrl'];
```
Comment (jakubtobiasz) [inline src/Payum/Action/Api/CreatePayByLinkTransactionAction.php:47]: Here should also be used the `\CommerceWeavers\SyliusTpayPlugin\Model\PaymentDetails` object

File: src/Tpay/Provider/TpayApiBankListProvider.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            $gateway->execute($value = new GetTpayTransactionsChannels(new ArrayObject()), true);
+
+            return $value;
+        });
```
Comment (jakubtobiasz) [inline src/Tpay/Provider/TpayApiBankListProvider.php:32]: As caching can be considered as a separate responsibility, and it makes it easier to decide whether we want to get cachable/non-cachable service, it is a good practice to use the decorator pattern. Ref:
- https://github.com/CommerceWeavers/SyliusTpayPlugin/blob/main/src/Tpay/Security/Notification/Resolver/CachedCertificateResolver.php

File: templates/shop/cart/complete/_payByLink.html.twig [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        border-color: blue;
+        background-color: #f0f8ff;
+    }
+
+    .bank-container {
+        display: grid;
+        grid-template-columns: repeat(4, 1fr);
+        gap: 15px;
+    }
+</style>
```
Comment (jakubtobiasz) [inline templates/shop/cart/complete/_payByLink.html.twig:29]: Create the `/assets/shop/scss/style.scss` and put the content there. Next, add the
```diff
+import './scss/style.scss';
```

to the `/assets/shop/entrypoint.js`

File: templates/shop/cart/complete/_payByLink.html.twig [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            tile.addEventListener('click', function () {
+                tiles.forEach(tileElement => tileElement.classList.remove('selected'));
+
+                tile.classList.add('selected');
+
+                hiddenInput.value = tile.getAttribute('data-bank-id');
+            });
+        });
+    });
+</script>
```
Comment (jakubtobiasz) [inline templates/shop/cart/complete/_payByLink.html.twig:57]: Follow the instruction above, adjusting the location of the file.

File: src/Tpay/Factory/CreatePayByLinkPayloadFactory.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function createFrom(PaymentInterface $payment, string $notifyUrl, string $localeCode): array
+    {
+        /** @var array{pay: array<string, mixed>} $payload */
+        $payload = $this->createRedirectBasedPaymentPayloadFactory->createFrom($payment, $notifyUrl, $localeCode);
+
+        /** @var array{tpay?: array{pay_by_link_channel_id?: string}} $paymentDetails */
+        $paymentDetails = $payment->getDetails();
```
Comment (jakubtobiasz) [inline src/Tpay/Factory/CreatePayByLinkPayloadFactory.php:25]: As previously, `PaymentDetails` object

Comment (jakubtobiasz) [reply src/Tpay/Factory/CreatePayByLinkPayloadFactory.php:26]: This object is our own, so you can just put what you need there :D

**Pull request merged**